spec/requests/users_controller_spec.rb

```diff
@@ -1,51 +1,2402 @@
+      end
+
+      it 'should fail if the user is old' do
+        # Older than the change period and >1 post
+        user.created_at = Time.now - (SiteSetting.username_change_period + 1).days
+        PostCreator.new(user,
+          title: 'This is a test topic',
+          raw: 'This is a test this is a test'
+        ).create
+
+        put "/u/#{user.username}/preferences/username.json", params: { new_username: new_username }
+
+        expect(response).to be_forbidden
+        expect(user.reload.username).to eq(old_username)
+      end
+
+      it 'should create a staff action log when a staff member changes the username' do
+        acting_user = Fabricate(:admin)
+        sign_in(acting_user)
+
+        put "/u/#{user.username}/preferences/username.json", params: { new_username: new_username }
+
+        expect(response).to be_success
+        expect(UserHistory.where(action: UserHistory.actions[:change_username], target_user_id: user.id, acting_user_id: acting_user.id)).to be_present
+        expect(user.reload.username).to eq(new_username)
+      end
+
+      it 'should return a JSON response with the updated username' do
+        put "/u/#{user.username}/preferences/username.json", params: { new_username: new_username }
+
+        expect(::JSON.parse(response.body)['username']).to eq(new_username)
+      end
+    end
+  end
+
+  describe '#check_username' do
+    it 'raises an error without any parameters' do
+      get "/u/check_username.json"
+      expect(response).not_to be_success
+    end
+
+    shared_examples 'when username is unavailable' do
+      it 'should return success' do
+        expect(response).to be_success
+      end
+
+      it 'should return available as false in the JSON' do
+        expect(::JSON.parse(response.body)['available']).to eq(false)
+      end
+
+      it 'should return a suggested username' do
+        expect(::JSON.parse(response.body)['suggestion']).to be_present
+      end
+    end
+
+    shared_examples 'when username is available' do
+      it 'should return success' do
+        expect(response).to be_success
+      end
+
+      it 'should return available in the JSON' do
+        expect(::JSON.parse(response.body)['available']).to eq(true)
+      end
+    end
+
+    it 'returns nothing when given an email param but no username' do
+      get "/u/check_username.json", params: { email: 'dood@example.com' }
+      expect(response).to be_success
+    end
+
+    context 'username is available' do
+      before do
+        get "/u/check_username.json", params: { username: 'BruceWayne' }
+      end
+      include_examples 'when username is available'
+    end
+
+    context 'username is unavailable' do
+      let!(:user) { Fabricate(:user) }
+      before do
+        get "/u/check_username.json", params: { username: user.username }
+      end
+      include_examples 'when username is unavailable'
+    end
+
+    shared_examples 'checking an invalid username' do
```

Comment:
We can just combine the assertions into a single `it` block otherwise we're running the setup block for each `it` block